### PR TITLE
Generic template collector factory

### PR DESCRIPTION
**Pull request by simonge**
### Briefly, what does this PR introduce?
Adds a template omni factory which creates a subset collection, combining the entries from an array of collections of the same type.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No

### Does this PR change default behavior?
No

### CODE REVIEW

File: src/algorithms/digi/CollectionCollector.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            const auto [in_collections] = input;
+            auto [out_collection]       = output;
+
+            out_collection->setSubsetCollection();
+
+            for (const auto& collection : in_collections) {
+                 for (const auto& hit : *collection) {
+                     out_collection->push_back(hit);
+                 }
+            //    out_collection->insert(out_collection->end(), collection->begin(), collection->end());
```
Comment (wdconinc) [inline src/algorithms/digi/CollectionCollector.h:46]:
```suggestion
```

File: src/factories/digi/CollectionCollector_factory.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    void Process(int64_t run_number, uint64_t event_number) {
+
+        try {
+            std::vector<gsl::not_null<const typename T::collection_type*>>  raw_collections;
+            for (const auto& in_collection : m_inputs()) {
+                raw_collections.push_back(gsl::not_null<const typename T::collection_type*>{in_collection});
+            }
+            typename T::collection_type* merged_collection = m_output().get();
+            m_algo->process(raw_collections,merged_collection);
```
Comment (wdconinc) [inline src/factories/digi/CollectionCollector_factory.h:38]:
```suggestion
            std::vector<gsl::not_null<const typename T::collection_type*>>  in_collections;
            for (const auto& in_collection : m_inputs()) {
                in_collections.push_back(gsl::not_null<const typename T::collection_type*>{in_collection});
            }
            typename T::collection_type* merged_collection = m_output().get();
            m_algo->process(in_collections, merged_collection);
```

File: src/global/tracking/tracking.cc [hunk 3]
```diff
@@ -17,7 +17,7 @@
 #include "TrackerMeasurementFromHits_factory.h"
 #include "extensions/jana/JChainMultifactoryGeneratorT.h"
 #include "extensions/jana/JOmniFactoryGeneratorT.h"
-#include "factories/tracking/TrackerHitCollector_factory.h"
+#include "factories/digi/CollectionCollector_factory.h"
```
Comment (wdconinc) [inline src/global/tracking/tracking.cc:20]: Might be useful to create a new type for this, `meta` or `generic`. It is as little a digi factory as a tracking factory.
```suggestion
#include "factories/meta/CollectionCollector_factory.h"
```

**Review by wdconinc — COMMENTED** [review]
Looks good.

**Comment by wdconinc** [comment]
This does seem to break all tracking, though. E.g. https://eicweb.phy.anl.gov/EIC/benchmarks/physics_benchmarks/-/jobs/2472823#L1479 for a downstream physics_benchmark. Same issue as in https://github.com/eic/EICrecon/pull/1215 so likely associated with the VariadicPodioInput somehow (@nathanwbrei).

**Comment by simonge** [comment]
> This does seem to break all tracking, though. E.g. https://eicweb.phy.anl.gov/EIC/benchmarks/physics_benchmarks/-/jobs/2472823#L1479 for a downstream physics_benchmark. Same issue as in #1215 so likely associated with the VariadicPodioInput somehow (@nathanwbrei).

I noticed the Capybara was unhappy and was going to try digging into it further but wasn't quite sure how. Everywhere else I've tested this it appears to work fine but then there's no other previous output to compare my other uses against.

**Comment by wdconinc** [comment]
> but then there's no other previous output to compare my other uses against

I usually download a `sim_....root` and corresponding `rec_....root` artifact from the latest run on main here, use the `sim_....root` as input for a local run of the branch inside eic-shell, then run `capybara bara --serve sim_...new.root sim_....root`.

File: src/factories/meta/CollectionCollector_factory.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        try {
+            std::vector<gsl::not_null<const typename T::collection_type*>>  in_collections;
+            for (const auto& in_collection : m_inputs()) {
+                in_collections.push_back(gsl::not_null<const typename T::collection_type*>{in_collection});
+            }
+            typename T::collection_type* merged_collection = m_output().get();
+            m_algo->process(in_collections, merged_collection);
+        }
+        catch(std::exception &e) {
+            throw JException(e.what());
```
Comment (wdconinc) [inline src/factories/meta/CollectionCollector_factory.h:41]: For debugging:
```suggestion
            this->logger()->error(e.what());
            throw JException(e.what());
```
I would not be surprised if this was a case of an exception getting silently eaten by JANA2, likely an exception in the `for` loop over `m_inputs()` that then results in not filling `merged_collection` at all.

Comment (simonge) [reply src/factories/meta/CollectionCollector_factory.h:41]: On my machine and events, for some currently undiagnosed reason running with this combination works:
SiBarrelTrackerRecHits
SiBarrelVertexRecHits
SiEndcapTrackerRecHits
TOFBarrelRecHit          
MPGDBarrelRecHits
MPGDDIRCRecHits       
B0TrackerRecHits

Adding TOFEndcapRecHits, starts calling the factory Process function then segfaults after a few events

Adding any of the below, the factory Process function is never called.
OuterMPGDBarrelRecHits
BackwardMPGDEndcapRecHits
ForwardMPGDEndcapRecHits

Comment (wdconinc) [reply src/factories/meta/CollectionCollector_factory.h:41]: While I hate suggesting it, for now we should probably just try-catch and print `e.what()` as a warning, check if that works in the scope of this PR, and dump the underlying issue on the plate of the TOF and MPGD group as an issue.

Comment (simonge) [reply src/factories/meta/CollectionCollector_factory.h:41]: I think this is a case for the optional collections, if any of them are missing the factory doesn't run. `e.what()` was never getting called when missing collections were requested.

It appears I was running on an old simulation file where different MPGD collections existed, looking at a new one the MPGD collections that broke my local run are present whereas MPGDDIRCHits doesn't appear to be. Can you confirm this?

Hopefully the TOF problems are related to the older simulation file too.

Comment (veprbl) [reply src/factories/meta/CollectionCollector_factory.h:41]: MPGD DIRC is for epic_arches, we don't need to support it anymore. I would remove `try catch`, if possible.

File: src/algorithms/meta/CollectionCollector.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+#include <algorithms/algorithm.h>
+#include <string>
+#include <string_view>
+
+#include "services/log/Log_service.h"
+#include "algorithms/interfaces/WithPodConfig.h"
+
+namespace eicrecon {
+
+   template<class T>
```
Comment (veprbl) [inline src/algorithms/meta/CollectionCollector.h:14]: Use two spaces for indent, please.

File: src/factories/meta/CollectionCollector_factory.h (hunk 4 above)
Comment (simonge) [reply src/factories/meta/CollectionCollector_factory.h:41]: If the reconstruction needs to work seamlessly with multiple detector configurations we'll need to allow for these to be optional variadic inputs. That does however allow the input list to get cluttered with unsupported collections. @nathanwbrei is this possible yet?

The list at the moment contains all of the detectors from a/b/c but will only work if they're all there. A separate issue but tangentially related, I hadn't realised the 18x275.xml was still a brycecanyon configuration and all of the CI is done using craterlake.

Comment (nathanwbrei) [reply src/factories/meta/CollectionCollector_factory.h:41]: "No hope" is overstating it quite a bit. Go ahead and merge what you have now, when I add optional collection support I'll update this at the same time.

**Review by wdconinc — APPROVED** [review]
No diffs. Thanks for this!

**Review by veprbl — APPROVED** [review]
Well done!

**Pull request merged**